File: TestLens/testlens-fixture/src/repositories/UserRepository.ts
import { User, UserId } from "../models/User";

export class UserRepository {
  private readonly usersById = new Map<UserId, User>();
  private readonly usersByEmail = new Map<string, UserId>();

  constructor(seedUsers: User[] = []) {
    for (const user of seedUsers) {
      this.save(user);
    }
  }

  save(user: User): void {
    this.usersById.set(user.id, user);
    this.usersByEmail.set(user.email, user.id);
  }

  findById(id: UserId): User | null {
    if (!id) {
      return null;
    }

    const user = this.usersById.get(id);
    if (!user) {
      return null;
    }

    return user;
  }

  findByEmail(email: string): User | null {
    if (!email) {
      return null;
    }

    const userId = this.usersByEmail.get(email);
    if (!userId) {
      return null;
    }

    return this.findById(userId);
  }

  delete(id: UserId): boolean {
    const existing = this.usersById.get(id);
    if (!existing) {
      return false;
    }

    this.usersById.delete(id);
    this.usersByEmail.delete(existing.email);
    return true;
  }
}

File: TestLens/testlens-fixture-rust/src/repositories/user_repository.rs
use crate::models::user::User;

#[derive(Debug, Default, Clone)]
pub struct UserRepository {
    users: Vec<User>,
}

impl UserRepository {
    pub fn new() -> Self {
        Self { users: Vec::new() }
    }

    pub fn save(&mut self, user: User) {
        self.users.push(user);
    }

    pub fn find_by_id(&self, id: u32) -> Option<User> {
        self.users.iter().find(|user| user.id == id).cloned()
    }

    pub fn find_by_email(&self, email: &str) -> Option<User> {
        self.users
            .iter()
            .find(|user| user.email.eq_ignore_ascii_case(email))
            .cloned()
    }
}
